enterprise/kernel/src/main/java/org/neo4j/kernel/impl/pagecache/WarmupAvailabilityListener.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.pagecache;
+
+import java.util.concurrent.TimeUnit;
+
+import org.neo4j.graphdb.factory.GraphDatabaseSettings;
+import org.neo4j.helpers.Format;
+import org.neo4j.kernel.AvailabilityGuard.AvailabilityListener;
+import org.neo4j.kernel.configuration.Config;
+import org.neo4j.logging.Log;
+import org.neo4j.scheduler.JobScheduler;
+
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+import static org.neo4j.scheduler.JobScheduler.Groups.pageCacheIOHelper;
+
+class WarmupAvailabilityListener implements AvailabilityListener
+{
+    private final JobScheduler scheduler;
+    private final PageCacheWarmer pageCacheWarmer;
+    private final Config config;
+    private final Log log;
+    private final PageCacheWarmerMonitor monitor;
+    private volatile JobScheduler.JobHandle profileHandle;
+
+    WarmupAvailabilityListener( JobScheduler scheduler, PageCacheWarmer pageCacheWarmer,
+                                Config config, Log log, PageCacheWarmerMonitor monitor )
+    {
+        this.scheduler = scheduler;
+        this.pageCacheWarmer = pageCacheWarmer;
+        this.config = config;
+        this.log = log;
+        this.monitor = monitor;
+    }
+
+    @Override
+    public void available()
+    {
+        scheduler.schedule( pageCacheIOHelper, this::startWarmup );
+    }
+
+    private void startWarmup()
+    {
+        try
+        {
+            long start = System.nanoTime();
+            pageCacheWarmer.reheat().ifPresent( pagesLoaded ->
+            {
+                long elapsedMillis = NANOSECONDS.toMillis( System.nanoTime() - start );
+                monitor.warmupCompleted( elapsedMillis, pagesLoaded );
+                log.debug( "Active page cache warmup took " + Format.duration( elapsedMillis ) +
+                           " to load " + pagesLoaded + " pages." );
```

Comment:
Would this timing logging be better within `PageCacheWarmer#reheat`? `PageCacheWarmerKernelExtension` is the only other place to call this and contains identical logging logic, with a different instance of `Monitor`.

Reply:
There's no logging in the kernel extension. I like that the `PageCacheWarmer` does not need to concern itself with any monitoring or logging.

Reply:
In that case this lambda, and that from line 93, could be extracted to be returned from a method.